### PR TITLE
[WIP] Adding better logic to handle training separate from inference

### LINKED ISSUE #150: Refactor: Remove duplicate training of som/w2v from inference function
**Is your feature request related to a problem? Please describe.**
@MichaelClifford @durandom We discussed that there is room for improvement in the code base and we do have training of the w2v and som happening twice. We also pull down data twice during a training run. I'd like to remove this duplicate effort and make some better design chooses.

In training function:
https://github.com/AICoE/log-anomaly-detector/blob/2d9ddfa7d11a79b0e350087889b756c5bcf055ce/anomaly_detector/anomaly_detector.py#L88-L114

In the inference function (To be removed):
https://github.com/AICoE/log-anomaly-detector/blob/2d9ddfa7d11a79b0e350087889b756c5bcf055ce/anomaly_detector/anomaly_detector.py#L142-L177

**Describe the solution you'd like**
I plan on removing the code in the inference function and have it do 1 thing only which is inference.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by zmhassan**
## Description
We have to refactor this code as it makes too many assumptions. Before we can fully refactor we need to remove duplicate code or parts that are not efficient. So when we train a model we want to prepare w2v and som model. The inference step should do only 1 thing. 

https://github.com/AICoE/log-anomaly-detector/blob/2d9ddfa7d11a79b0e350087889b756c5bcf055ce/anomaly_detector/anomaly_detector.py#L142-L177

# Fixes
#150

### CODE REVIEW

**Comment by zmhassan** [comment]
@durandom This is the first wave of refactoring of LAD. Please let me know if this PR is good. if so then please merge.

**Comment by zmhassan** [comment]
@fridex @goern is thoth bot broken? I'm seeing the same module error again even though everything passes in travis. Also coala passes locally when I test.

File: anomaly_detector/anomaly_detector.py [hunk 1]
```diff
@@ -149,86 +154,53 @@ def infer(self, false_positives=None):
...
+        f = []
+        _LOGGER.info("Max anomaly score: %f" % max(dist))
+        for i in range(len(data)):
+            _LOGGER.debug("Updating entry %d - dist: %f; mean: %f" % (i, dist[i], mean))
+
+            # TODO: This needs to be more general,
+            #       only works for ES incoming logs right now.
+            s = json_logs[i]
+            s["predict_id"] = str(uuid.uuid4())
+            s["anomaly_score"] = dist[i]
```
Comment (MichaelClifford) [inline anomaly_detector/anomaly_detector.py]: @zmhassan This approach is a little bit confusing to me. I see that you are setting the anomaly score from the `dist[ ]` list. But you are providing that list as an argument to the `infer()` function. The way I think about it is like this: computing the `dist[ ]` values and then checking whether they are higher or lower than the threshold is the inference step. And I don't see where those distance values are being calculated for new data. I only see these values being calculated in the the `train()` function, so its unclear to me where the inference on new data is happening. Let me know if I am mistaken, but this reads to me like training and inference is being performed on the same data. Let me know if I've overlooked something : ) thanks.

**Comment by zmhassan** [comment]
stream

File: anomaly_detector/anomaly_detector.py (hunk 1 above)
Comment (zmhassan) [reply anomaly_detector/anomaly_detector.py]: That is correct. I am using the same data that was used to train to do the inference. 

https://github.com/AICoE/log-anomaly-detector/blob/4e8834da9f96409d01fee6f9b4cdf762f23e490d/anomaly_detector/anomaly_detector.py#L221-L238

In the run function you'll see I pass that in and we also pass in the false positives as well.

Comment (zmhassan) [reply anomaly_detector/anomaly_detector.py]: ```             self.infer(false_positives=false_positives, dist=dist, data=data, json_logs=json_logs) ```

**Comment by goern** [comment]
recheck

**Comment by zmhassan** [comment]
@goern thank you 👍

**Comment by zmhassan** [comment]
@durandom I got a meeting with Michael tomorrow afternoon. Let's put this PR on hold. I'll message you once it's ready to land.

**Comment by zmhassan** [comment]
Closing this PR in favor of PR: https://github.com/AICoE/log-anomaly-detector/pull/153.

**Pull request closed without merging**